### PR TITLE
🐛Set default UI for amp-sidebar within amp-story

### LINKED ISSUE #19423: Set default UI for amp-sidebar within amp-story
Currently, `<amp-sidebar>` components within `<amp-story>` have a default gray background color, a non-standard animation, and can be of any width.

Ideally, we would want to default the background color to white, and enforce a specific width and opening/closing animation.

Part of #16532

### PR DESCRIPTION
**Pull request by bramanudom**
resolves #19423

The default UI for the amp-sidebar within amp-story 
- Has a default background color of white
- Has specific widths set for different screen resolutions (mobile, desktop)
- Has a standard transition. 

A visual test was also added.

### CODE REVIEW

File: extensions/amp-story/1.0/amp-story.css [hunk 1]
```diff
@@ -176,6 +176,22 @@ amp-story-page[active] {
   z-index: 1 !important;
 }
 
+amp-sidebar {
+  background-color: #ffffff !important;
+  max-width: 280px !important;
+  transition: opacity 0.3s cubic-bezier(0.0, 0.0, 0.2, 1) !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: Can we do a slide in/out transition instead, similar to what we have for the share menu?

**Comment by bramanudom** [comment]
As we talked about, the amp-sidebar does create a mask when the sidebar is toggled to open. The mask is a div that's just appended to the document body. Are there any thoughts of what might be the best way to handle this?

It looks like on the javascript side of amp-story we are already observing the sidebar, would it make sense to consider removing the mask element out of the DOM?

**Comment by gmajoulet** [comment]
Our goal of having the mask between the amp-story and the sidebar is made impossible by the fact that the mask is a sibling of amp-story, and the sidebar a child of amp-story:
```
<body>
  <amp-story>
    <amp-sidebar></amp-sidebar>
  </amp-story>
  <mask></mask>
</body>
```

One solution could be to:
- reuse the code to add an opacity overlay to the story ([here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/1.0/amp-story.css#L212-L223))
- hide the `.i-amphtml-sidebar-mask` element
- listen for click events on our own layer, and close the sidebar when it happens

Wdyt?

**Comment by gmajoulet** [comment]
Ah, sorry for the confusion... The snippet I wrote above was showing what's happening right now, and why we have to change this. The issue you just mentioned of having the mask on top of the sidebar is problematic for many reasons: it's going to darken all the content, and prevents from clicking on elements in the sidebar (unless you use `pointer-events: none;`, but then you would not be able to close the sidebar when the user clicks on the mask, so this doesn't work either).

What we have right now doesn't allow the mask to be on top of the story, and under the sidebar, leading to the issue you mentioned:

```
<body>
  <amp-story>
    <amp-sidebar></amp-sidebar>
  </amp-story>
  <mask></mask>
</body>
```

What we want to have, and could be achieved with the solution I mentioned above (basically it means hiding the default amp-sidebar mask, and using our own):

```
<body>
  <amp-story>
    <mask></mask>
    <amp-sidebar></amp-sidebar>
  </amp-story>
</body>
```

File: examples/visual-tests/amp-story/amp-story-sidebar.html [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      .i-amphtml-story-spinner-container {
+        /* Hide the spinner for loading pages, since this causes flakes */
+        display: none;
+      }
+      h1.hello-world {
+        color: white;
+        text-shadow: 3px 3px rgba(0, 0, 0, 0.5);
+      }
+    </style>
+    <script>
```
Comment (gmajoulet) [inline examples/visual-tests/amp-story/amp-story-sidebar.html]: Nit: looks like you can remove this <script> block. #17108 looks fixed, and you're already navigating to the second page using the new API in the other file :)

File: extensions/amp-story/1.0/amp-story.js [hunk 3]
```diff
@@ -1455,6 +1460,39 @@ export class AmpStory extends AMP.BaseElement {
     }
   }
 
+  /**
+   * @private
+   */
+  openMask_() {
+    const actions = Services.actionServiceForDoc(this.getAmpDoc());
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Nit: please move this inside the click listener, so we only get the service when we need it

File: extensions/amp-story/1.0/amp-story.js [hunk 4]
```diff
@@ -1455,6 +1460,39 @@ export class AmpStory extends AMP.BaseElement {
...
+    if (!this.maskElement_) {
+      const mask = this.win.document.createElement('div');
+      mask.classList.add('i-amphtml-amp-story-sidebar-mask');
+      mask.addEventListener('click', () => {
+      actions.execute(this.sidebar_, 'close', /* args */ null,
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Nit: indentation

File: extensions/amp-story/1.0/amp-story.js [hunk 5]
```diff
@@ -1455,6 +1460,39 @@ export class AmpStory extends AMP.BaseElement {
...
   }
 
+  /**
+   * @private
+   */
+  openMask_() {
+    const actions = Services.actionServiceForDoc(this.getAmpDoc());
+    if (!this.maskElement_) {
+      const mask = this.win.document.createElement('div');
+      mask.classList.add('i-amphtml-amp-story-sidebar-mask');
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Nit: `.i-amphtml-story-opacity-mask` (we don't add the `amp-` prefix to `story`, and we might want to re use the mask for other features : )

File: extensions/amp-story/1.0/amp-story.js [hunk 6]
```diff
@@ -1455,6 +1460,39 @@ export class AmpStory extends AMP.BaseElement {
...
+      mask.addEventListener('click', () => {
+      actions.execute(this.sidebar_, 'close', /* args */ null,
+        /* source */ null, /* caller */ null, /* event */ null,
+            ActionTrust.HIGH);
+      });
+      // The mask div should be added right after the sidebar.
+      const referenceNode = this.win.document.getElementsByTagName('amp-sidebar')[0];
+      referenceNode.parentNode.insertBefore(mask, referenceNode.nextSibling);
+      mask.addEventListener('touchmove', e => {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Why is this needed? If you're trying to prevent other things from happening, like the "tap next" navigation UI help, there are better ways to do it, like adding the `SIDEBAR_STATE` to this list: https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/1.0/amp-story.js#L611
Lmk if you had to use this for another reason that I missed :)

File: extensions/amp-story/1.0/amp-story.js [hunk 7]
```diff
@@ -1455,6 +1460,39 @@ export class AmpStory extends AMP.BaseElement {
...
+      const mask = this.win.document.createElement('div');
+      mask.classList.add('i-amphtml-amp-story-sidebar-mask');
+      mask.addEventListener('click', () => {
+      actions.execute(this.sidebar_, 'close', /* args */ null,
+        /* source */ null, /* caller */ null, /* event */ null,
+            ActionTrust.HIGH);
+      });
+      // The mask div should be added right after the sidebar.
+      const referenceNode = this.win.document.getElementsByTagName('amp-sidebar')[0];
+      referenceNode.parentNode.insertBefore(mask, referenceNode.nextSibling);
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Given the `z-index` that you set, you should be able to append the mask at any position as long as it's a direct child of `amp-story` (which you can refer to as `this.element` since you're in the amp-story component).
I believe you could do: `this.element.appendChild(maskEl)`

File: extensions/amp-story/1.0/amp-story.js [hunk 8]
```diff
@@ -1455,6 +1460,39 @@ export class AmpStory extends AMP.BaseElement {
     }
   }
 
+  /**
+   * @private
+   */
+  openMask_() {
+    const actions = Services.actionServiceForDoc(this.getAmpDoc());
+    if (!this.maskElement_) {
+      const mask = this.win.document.createElement('div');
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Nit: `maskElement` or `maskEl` (the styleguide recommends not using abbreviations, but `El` is a common exception to this rule)

File: extensions/amp-story/1.0/amp-story.js [hunk 9]
```diff
@@ -1444,7 +1447,9 @@ export class AmpStory extends AMP.BaseElement {
         actions.execute(this.sidebar_, 'open', /* args */ null,
         /* source */ null, /* caller */ null, /* event */ null,
             ActionTrust.HIGH);
+        this.openMask_()
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Nit: since this file is unfortunately a 2k+ lines monster with a lot of different things, what do you think of adding something like `opacity` in the same, ie `openOpacityMask`?

File: extensions/amp-story/1.0/amp-story.css [hunk 10]
```diff
@@ -232,6 +243,24 @@ amp-story-page[active] {
   }
 }
 
+.i-amphtml-amp-story-sidebar-mask {
+  position: fixed !important;
+  top: 0 !important;
+  left: 0 !important;
+  width: 100vw !important;
+  height: 100vh !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: Since the sidebar is a direct child of `amp-story`, you can use `position: absolute; top: 0; bottom: 0; width: 100%; height: 100%;`. We try to position everything after `amp-story`, for if we ever need to display several stories in the same page.

Also, please add some `!important` statements at the end of all the CSS rules, so publishers can't override them :))

File: extensions/amp-story/1.0/amp-story.js (hunk 9 above)
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story.js]: Makes perfect sense to me!

File: extensions/amp-story/1.0/amp-story.js [hunk 11]
```diff
@@ -1455,6 +1461,36 @@ export class AmpStory extends AMP.BaseElement {
...
+    }
+    toggle(this.maskElement_, /* display */true);
+  }
+
+  /**
+   * @private
+   */
+  closeOpacityMask_() {
+    if (this.maskElement_) {
+      toggle(this.maskElement_, /* display */false);
```
[the remark this answers is not in the document]
Comment (cathyxz) [reply extensions/amp-story/1.0/amp-story.js]: I think the correct API for this is now `this.mutateElement` because we have deprecated `vsync` everywhere. In fact stories is one of the last extensions still on it.

**Comment by bramanudom** [comment]
I pushed the changes that have been made addressing both @gmajoulet 's online and offline comments. 

It should be noted that there are still issues regarding the in/out transitions for the sidebar, but the issues seem to be connected to an existing issue in amp-sidebar. See @gmajoulet 's filed ticket [here](https://github.com/ampproject/amphtml/issues/19777).

File: extensions/amp-story/1.0/amp-story.css [hunk 12]
```diff
@@ -176,6 +176,26 @@ amp-story-page[active] {
...
 
+amp-story amp-sidebar[side][open] {
+  background-color: #ffffff !important;
+  max-width: 280px !important;
+  min-width: 280px !important;
+  width: 280px !important;
+  display: block !important;
+}
+
+amp-story amp-sidebar[side=right] {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: Do we need to duplicate this block?

File: extensions/amp-story/1.0/amp-story.css [hunk 13]
```diff
@@ -176,6 +176,26 @@ amp-story-page[active] {
...
+  max-width: 280px !important;
+  min-width: 280px !important;
+  width: 280px !important;
+  display: block !important;
+}
+
+[desktop] amp-sidebar[side][open],
+[desktop] amp-sidebar[side=right] {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: Same here, do we need both selectors?

File: extensions/amp-story/1.0/amp-story.css [hunk 14]
```diff
@@ -232,6 +252,43 @@ amp-story-page[active] {
...
+  top: 0 !important;
+  bottom: 0 !important;
+  width: 100% !important;
+  height: 100% !important;
+  opacity: 0.55 !important;
+  /* Prevent someone from making this a full-screen image */
+  background-image: none !important;
+  background-color: #000 !important;
+  z-index: 100001 !important;
+  animation: i-amphtml-bookend-overlay-opacity 0.2s cubic-bezier(0.4, 0.0, 1, 1) forwards !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: You could actually do
`transition: opacity .... !important` (and remove the `forwards`), that'd allow you to remove the whole `@keyframes i-amphtml-bookend-overlay-opacity-hidden` block.

Also, it looks like you added the closing animation here, when this is actually the visible class. Could you please switch both animations duration/curves?

File: extensions/amp-story/1.0/amp-story.css [hunk 15]
```diff
@@ -232,6 +252,43 @@ amp-story-page[active] {
...
+}
+
+.i-amphtml-story-opacity-mask-hidden {
+  background-color: #000 !important;
+  opacity: 0 !important;
+  background-image: none !important;
+  pointer-events: none !important;
+  z-index: 100001 !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: Nit: you should not need to repeat the `z-index` and the `background` statements

File: extensions/amp-story/1.0/amp-story.js [hunk 16]
```diff
@@ -164,6 +164,20 @@ const MINIMUM_AD_MEDIA_ELEMENTS = 2;
...
+ * @const {string}
+ */
+const OPACITY_MASK_CLASS_NAME = 'i-amphtml-story-opacity-mask';
+
+/**
+ * CSS class for the opacity layer that separates the amp-sidebar and the rest
+ * of the story when the amp-sidebar is leaving the screen
+ * @const {string}
+ */
+const OPACITY_MASK_HIDDEN_CLASS_NAME = 'i-amphtml-story-opacity-mask-hidden';
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Can we add a `[hidden]` attribute on `OPACITY_MASK_CLASS_NAME` instead, so we don't have to declare two classes?
`.i-amphtml-story-opacity-mask[hidden] {/* opacity + transition + pointer-events overrides */}`

File: extensions/amp-story/1.0/amp-story.js [hunk 17]
```diff
@@ -1487,19 +1505,59 @@ export class AmpStory extends AMP.BaseElement {
...
+      this.openOpacityMask_();
       this.storeService_.dispatch(Action.TOGGLE_SIDEBAR, false);
     }
   }
 
+  /**
+   * @private
+   */
+  openOpacityMask_() {
+    this.mutateElement(() => {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Nit: only the parts where you write in the DOM should be wrapped in this `mutateElement`: the classList.add/remove, and the appendChild

File: extensions/amp-story/1.0/amp-story.css (hunk 12 above)
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story.css]: Ah, thanks for the catch. This was originally added for setting transitions for the the sidebar closing. Now that we are decoupling those changes with the mask/background changes, I will remove this.

**Comment by bramanudom** [comment]
PTAL at the changes when you have time ! @newmuis @gmajoulet @cvializ

File: extensions/amp-story/1.0/amp-story.js [hunk 18]
```diff
@@ -1487,19 +1498,54 @@ export class AmpStory extends AMP.BaseElement {
...
+              ActionTrust.HIGH);
+          this.closeOpacityMask_();
+        }
+      });
+      this.maskElement_ = maskEl;
+      this.mutateElement(() => {
+        this.element.appendChild(this.maskElement_);
+      });
+    }
+    toggle(this.maskElement_, /* display */true);
```
[the remark this answers is not in the document]
Comment (newmuis) [reply extensions/amp-story/1.0/amp-story.js]: As in, specify content in the sidebar by default?  That's a non-goal

File: extensions/amp-story/1.0/amp-story.js [hunk 19]
```diff
@@ -1487,19 +1498,58 @@ export class AmpStory extends AMP.BaseElement {
...
+    });
+  }
+
+  /**
+   * @private
+   */
+  closeOpacityMask_() {
+    if (this.maskElement_) {
+      this.mutateElement(() => {
+        toggle(this.maskElement_, /* display */false);
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Here and above the compiler is complaining that `this.maskElement_` could be null... :(( You can work around that by passing `dev().assertElement(this.maskElement_)` as the first parameter of `toggle`

File: extensions/amp-story/1.0/amp-story.js [hunk 20]
```diff
@@ -1487,19 +1498,58 @@ export class AmpStory extends AMP.BaseElement {
...
+  /**
+   * @private
+   */
+  openOpacityMask_() {
+    if (!this.maskElement_) {
+      const maskEl = this.win.document.createElement('div');
+      maskEl.classList.add(OPACITY_MASK_CLASS_NAME);
+      maskEl.addEventListener('click', () => {
+        const actions = Services.actionServiceForDoc(this.element);
+        if (this.sidebar_) {
```
Comment (cvializ) [inline extensions/amp-story/1.0/amp-story.js]: Was this `if` to make Closure Compiler happy? It looks like `openOpacityMask_` and `closeOpacityMask_` is only ever called from a function that already check `if (this.sidebar_)` so the check is redundant:

> https://github.com/ampproject/amphtml/blob/27bf8ae175dbf55a0fadc59a55a7432bb3fff3a8/extensions/amp-story/1.0/amp-story.js#L1498-L1503
`openOpacityMask_` and its calling function both check `if (this.sidebar_)` and only one should.

You can avoid redundant checks a couple of ways.

1. Add a non-nullable `!Element` parameter to the method and pass the reference when you call `openOpacityMask_` is called.
2. Use `dev().assertElement(this.sidebar_)` when you just need to make Closure Compiler happy. These calls are stripped at compile-time to avoid putting them in the final binary sent to the client.

This will make it super clear in the code when `this.sidebar_` is and isn't supposed to be defined.

Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story.js]: Thank you for catching this redundancy! The check was added in an earlier iteration, and I actually think that we should just be able to get rid of it entirely. I'll take notes on the different ways I can avoid redundancy for future PRs.
